### ISSUE TITLE
Question about closing current view in a sequence

### ISSUE THREAD

**Issue by igordeoliveirasa** [uber/RIBs#156]
Imagine this sequence:
Root(owns view) -> A(not owns view) -> B(not owns view) -> C(owns view).

If CViewController is closed by itself, I'd like to back to Root state.

So, on CViewController, I have a button to close current view. 
I suppose that this button would call:

```swift
self.listener.didCloseTap()
```

I can imagine that this listener is B Interactor. However, what should B interactor do?

I don't know how to come back.

Could somebody help me with this?

P.S.: Sorry about opening multiple questions.. It's because I really need this in a short time.

Thanks.

**Comment by sbarow** [comment]
You would propagate that listener message all the way back to root in this case.

i.e `Root(owns view)` <- `self.listener.myMethodOtherForClose()` `A(not owns view)` <- `self.listener.myMethodForClose()` `B(not owns view)` <- `self.listener.didCloseTap()` `C(owns view)`

**Comment by igordeoliveirasa** [comment]
@sbarow thank you by you answer!
But, root does not have access to dismiss B view controller...

**Comment by sbarow** [comment]
You can choose who dismisses the view controller, which can either be done by C itself or by B, the message i.e the `self.listener.myMethodForClose()` can still be called letting A know something has happened.

**Comment by igordeoliveirasa** [comment]
@sbarow
We know that B has attached C... So, B is the responsible for detaching and dismissing C. I need something more concrete...

Are the next steps possible AND good approaches?
- CViewController calls its self.listener.didCCloseTap()
- self.listener is B Interactor
- So, B Interactor calls router.detach()
- Router detaches current child and dismiss C ViewController and calls self.iteractor.listener.didBClose()
- Since self.iteractor.listener is A Interactor, A Interactor calls: self.router.detach()
- self.router is A Router. It detaches current child and calls: self.interactor.listener.didAClose()
- Since self.iteractor.listener is Root Interactor, Root Interactor calls: self.router.detach()
- self.router is Root Router. It detaches current child stop there.

**Comment by igordeoliveirasa** [comment]
@sbarow it won't help. Since router detach method is able to detach current view controller but has not business inteligence/logic to call next parent interactor.

I really need a clear instruction about this.

**Comment by sbarow** [comment]
- "So, B Interactor calls router.detach()" - and calls listener (from interactor) `So, B Interactor calls self.listener.didBClose()`

Same pattern follows.

**Comment by igordeoliveirasa** [comment]
@sbarow 
> Same pattern follows.
It's impossible. Since router.detach() knows that it should detach current view, however, it does not know who has called him so as to decide that it should now call self.interactor.listener.didBClose() ...

**Comment by sbarow** [comment]
I am not sure I follow. B router attached C router right?

If so:
- C view controller calls listener (C interactor ) to say its done i.e back/close button tapped.
- C interactor calls listener (B interactor) to say C is done, decide what to do.
- B Interactor tells B router detach C router.
- B Interactor tells listener (A Interactor) to say B is done, decide what to do.
- A interactor tells A router detach B router.
- A interactor tells listener (Root interactor) to say A is done, decide what to do.
- Root decides what to do.

Make sense?

**Comment by igordeoliveirasa** [comment]
> B Interactor tells B router detach C router.
> B Interactor tells listener (A Interactor) to say B is done, decide what to do.

Two commands made by B Interactor, right? First informing router to detach and after informing its listener that it needs to be detached...

Did I understand you?

**Comment by sbarow** [comment]
Yip, the thing to remember is the interactor is always responsible for making logic decisions wether that be tell the router to detach or telling its listener i.e its parent interactor that it needs to make a decision based on some thing.

**Comment by igordeoliveirasa** [comment]
@sbarow AWESOME! I got it.
I guess it would be more safe If B Router detach's method could offer a completion callback, am I right? Since there is a animation dismissing viewcontroller, It's better to execute second command (which is: letting A Interactor to be informed) after dismissing completion.

Am I right, @sbarow ?

**Comment by sbarow** [comment]
Yeah, I updated my previous comment with that info :-)

**Comment by igordeoliveirasa** [comment]
@sbarow Thank you so much! Now I'm feeling well again. lol
